lib/open_api_spex.ex

```diff
@@ -115,12 +147,67 @@ defmodule OpenApiSpex do
   """
   defmacro schema(body) do
     quote do
+      @compile {:report_warnings, false}
       @behaviour OpenApiSpex.Schema
-      @schema struct(OpenApiSpex.Schema, Map.put(unquote(body), :"x-struct",  __MODULE__))
+      @schema struct(OpenApiSpex.Schema, Map.put(unquote(body), :"x-struct", __MODULE__))
       def schema, do: @schema
       @derive [Poison.Encoder]
       defstruct Schema.properties(@schema)
       @type t :: %__MODULE__{}
+
+      Map.from_struct(@schema) |> OpenApiSpex.validate_compiled_schema()
     end
   end
+
+  @doc """
+  Validate the compiled schema's properties to ensure the schema is not improperly
+  defined. Only errors which would cause a given schema to _always_ fail should be
+  raised here.
+  """
+  def validate_compiled_schema(schema) do
+    Enum.each(schema, fn prop_and_val ->
+      :ok = validate_compiled_schema(prop_and_val, schema)
+    end)
+  end
+
+  def validate_compiled_schema({_, %Schema{} = schema}, _parent) do
+    validate_compiled_schema(schema)
+  end
+
+  @doc """
+  Used for validating the schema at compile time, otherwise we're forced
+  to raise errors for improperly defined schemas at runtime.
+  """
+  def validate_compiled_schema({:discriminator, %{propertyName: property, mapping: _}}, %{
```

Comment:
If this is for internal use only then why not make it private function?

Reply:
@hauleth we can totally make this private if you don't see value in keeping it public

Reply:
This function is called as a public function, unless I'm misinterpreting the call:

https://github.com/open-api-spex/open_api_spex/pull/69/files/c31a292d8d82e06a8d98c9abbd410f46af07abaa#diff-355a2bb78931a9e2c80a161db03cbb8dR158

Reply:
@moxley yes, as the code is currently written that is a public function and needs to be publicly accessible because it's being called from within the context of a macro. We can change that but it would be a larger undertaking and based on the potential move to `ex_json_schema` I wouldn't recommend making the change unless the maintainers require the change for merging.